### PR TITLE
Add pull secret management for DTK

### PR DESCRIPTION
**Pull request by pacevedom**
When deploying in GCP the crane package gets into an infinite loop if traffic towards metadata server is dropped. This makes SRO unresponsive.
Changing authn discovery from crane package into a manual approach where we grab the pull secrets from a secret will avoid the use of the metadata server while still being able to authenticate and obtain the same info.

Signed-off-by: Pablo Acevedo Montserrat <pacevedo@redhat.com>

### CODE REVIEW

File: pkg/registry/registry.go [hunk 1]
```diff
@@ -39,7 +38,20 @@ type DriverToolkitEntry struct {
...
 }
 
+func writeImageRegistryCredentials() {
+	secretNamespace := "openshift-config"
+	secretName := "pull-secret"
+	configFilePath := "/home/nonroot/.docker/config.json"
+	secretDataName := ".dockerconfigjson"
+
+	s, err := clients.Interface.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	exit.OnError(err)
```
Comment (dagrayvid) [inline pkg/registry/registry.go]: Instead of exit.OnError, can we log the error (missing global pull-secret) and keep trying to reconcile?

Comment (pacevedom) [reply pkg/registry/registry.go]: Just realised I lost the vanilla k8s check, so I added that back. About logging and trying to reconcile, this is used from `LastLayer` function, which bails out in almost every error, so I will log and return errors and let the caller decide.

**Comment by SchSeba** [comment]
Hi @pacevedom quick question why do you need to create an empty folder in the pod definition?

**Comment by pacevedom** [comment]
@SchSeba the crane package will look for existing credentials (if not configured otherwise, but this was the source of the issue in the first place) in `$HOME/.docker/config.json`. Since the pod is using a read-only root file system we need to create that dir so that its writable, hence the memory-backed volume.

**Comment by SchSeba** [comment]
Oh sorry I miss that the pod file system is RO.

Thanks!
/lgtm

File: pkg/registry/registry.go [hunk 2]
```diff
@@ -39,7 +38,33 @@ type DriverToolkitEntry struct {
 	OSVersion           string `json:"OSVersion"`
 }
 
+func writeImageRegistryCredentials() error {
+	secretNamespace := "openshift-config"
```
Comment (SchSeba) [inline pkg/registry/registry.go]: nit: you can have this a const so we don't need to allocate memory

File: pkg/registry/registry.go [hunk 3]
```diff
@@ -39,7 +38,33 @@ type DriverToolkitEntry struct {
...
+		log.Info("Can not find namespace for pull secrets, assuming vanilla k8s")
+		return nil
+	}
+
+	s, err := clients.Interface.CoreV1().Secrets(secretNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Can not retrieve pull secrets")
+	}
+
+	err = os.WriteFile(configFilePath, s.Data[secretDataName], 0644)
```
Comment (SchSeba) [inline pkg/registry/registry.go]: please add a check that the secretDataName exist in the configmap

**Review by SchSeba — COMMENTED** [review]
2 small comments

File: pkg/registry/registry.go [hunk 4]
```diff
@@ -39,7 +45,35 @@ type DriverToolkitEntry struct {
 	OSVersion           string `json:"OSVersion"`
 }
 
+func writeImageRegistryCredentials() error {
+	configFilePath := "/home/nonroot/.docker/config.json"
```
Comment (SchSeba) [inline pkg/registry/registry.go]: duplicate with https://github.com/openshift-psap/special-resource-operator/pull/110/files#diff-cb61d8f79451b9541de4a8cc0811523a68d15452b2f5971c7618ea5b423cf4ecR30?

**Review by SchSeba — COMMENTED** [review]
great just remove the duplicate variable and this PR is ready from my point of view

File: pkg/registry/registry.go (hunk 4 above)
Comment (pacevedom) [reply pkg/registry/registry.go]: oops, a leftover. Thanks!

**Review by SchSeba — COMMENTED** [review]
lgtm

**Comment by SchSeba** [comment]
Oh @pacevedom you should  also update this file for the bundle

https://github.com/openshift-psap/special-resource-operator/blob/master/bundle/4.9/manifests/special-resource-operator.v4.9.0.clusterserviceversion.yaml

**Pull request merged**